File: connector/PlayHouse.Connector/Connector.cs
#nullable enable

using PlayHouse.Connector.Internal;
using PlayHouse.Connector.Protocol;

namespace PlayHouse.Connector;

/// <summary>
/// PlayHouse 클라이언트 Connector
/// </summary>
/// <remarks>
/// 클라이언트가 Play Server에 연결하여 실시간 통신을 수행하는 메인 클래스입니다.
/// serviceId 파라미터가 제거되어 하나의 연결에 하나의 서비스만 사용합니다.
/// </remarks>
public sealed class Connector : IConnectorCallback
{
    private ClientNetwork? _clientNetwork;
    private bool _disconnectFromClient;

    /// <summary>
    /// Connector 설정
    /// </summary>
    public ConnectorConfig ConnectorConfig { get; private set; } = new();

    #region Events

    /// <summary>
    /// 연결 결과 이벤트
    /// </summary>
    public event Action<bool>? OnConnect;

    /// <summary>
    /// 메시지 수신 이벤트 (stageId, packet)
    /// stageId가 0이면 Stage 없는 메시지
    /// </summary>
    public event Action<long, IPacket>? OnReceive;

    /// <summary>
    /// 에러 이벤트 (stageId, errorCode, request)
    /// </summary>
    public event Action<long, ushort, IPacket>? OnError;

    /// <summary>
    /// 연결 끊김 이벤트
    /// </summary>
    public event Action? OnDisconnect;

    #endregion

    #region IConnectorCallback Implementation

    void IConnectorCallback.ConnectCallback(bool result)
    {
        OnConnect?.Invoke(result);
    }

    void IConnectorCallback.ReceiveCallback(long stageId, IPacket packet)
    {
        OnReceive?.Invoke(stageId, packet);
    }

    void IConnectorCallback.ErrorCallback(long stageId, ushort errorCode, IPacket request)
    {
        OnError?.Invoke(stageId, errorCode, request);
    }

    void IConnectorCallback.DisconnectCallback()
    {
        if (_disconnectFromClient)
        {
            return;
        }

        OnDisconnect?.Invoke();
    }

    #endregion

    #region Initialization

    /// <summary>
    /// Connector 초기화
    /// </summary>
    /// <param name="config">설정</param>
    public void Init(ConnectorConfig config)
    {
        ConnectorConfig = config ?? throw new ArgumentNullException(nameof(config));
        _clientNetwork = new ClientNetwork(config, this);
    }

    #endregion

    #region Connection

    /// <summary>
    /// 서버에 연결 (비동기로 OnConnect 이벤트 발생)
    /// </summary>
    /// <param name="debugMode">디버그 모드</param>
    public void Connect(bool debugMode = false)
    {
        _disconnectFromClient = false;
        _clientNetwork!.Connect(debugMode);
    }

    /// <summary>
    /// 서버에 비동기 연결
    /// </summary>
    /// <param name="debugMode">디버그 모드</param>
    /// <returns>연결 성공 여부</returns>
    public async Task<bool> ConnectAsync(bool debugMode = false)
    {
        _disconnectFromClient = false;
        return await _clientNetwork!.ConnectAsync(debugMode);
    }

    /// <summary>
    /// 서버 연결 끊기
    /// </summary>
    public void Disconnect()
    {
        _disconnectFromClient = true;
        _ = _clientNetwork?.DisconnectAsync();
    }

    /// <summary>
    /// 연결 상태 확인
    /// </summary>
    public bool IsConnected()
    {
        return _clientNetwork?.IsConnect() ?? false;
    }

    /// <summary>
    /// 인증 상태 확인
    /// </summary>
    public bool IsAuthenticated()
    {
        return _clientNetwork?.IsAuthenticated() ?? false;
    }

    #endregion

    #region Authentication

    /// <summary>
    /// 인증 메시지 ID 설정
    /// 등록된 메시지만 인증 전에 전송 가능
    /// </summary>
    /// <param name="msgId">인증 메시지 ID</param>
    public void SetAuthenticateMessageId(string msgId)
    {
        _clientNetwork?.SetAuthenticateMessageId(msgId);
    }

    /// <summary>
    /// 인증 요청 (콜백 방식)
    /// </summary>
    /// <param name="request">인증 요청 패킷</param>
    /// <param name="callback">응답 콜백</param>
    public void Authenticate(IPacket request, Action<IPacket> callback)
    {
        if (!IsConnected())
        {
            OnError?.Invoke(0, (ushort)ConnectorErrorCode.Disconnected, request);
            return;
        }

        _clientNetwork!.Request(request, callback, 0, isAuthenticate: true);
    }

    /// <summary>
    /// 인증 요청 (async/await 방식)
    /// </summary>
    /// <param name="request">인증 요청 패킷</param>
    /// <returns>인증 응답 패킷</returns>
    public async Task<IPacket> AuthenticateAsync(IPacket request)
    {
        if (!IsConnected())
        {
            throw new ConnectorException(0, (ushort)ConnectorErrorCode.Disconnected, request, 0);
        }

        return await _clientNetwork!.RequestAsync(request, 0, isAuthenticate: true);
    }

    #endregion

    #region Send/Request (No Stage)

    /// <summary>
    /// 메시지 전송 (Stage 없음, 응답 없음)
    /// </summary>
    /// <param name="packet">전송할 패킷</param>
    public void Send(IPacket packet)
    {
        if (!IsConnected())
        {
            OnError?.Invoke(0, (ushort)ConnectorErrorCode.Disconnected, packet);
            return;
        }

        if (!IsAuthenticated())
        {
            OnError?.Invoke(0, (ushort)ConnectorErrorCode.Unauthenticated, packet);
            return;
        }

        _clientNetwork!.Send(packet, 0);
    }

    /// <summary>
    /// 요청 전송 (Stage 없음, 콜백 방식)
    /// </summary>
    /// <param name="request">요청 패킷</param>
    /// <param name="callback">응답 콜백</param>
    public void Request(IPacket request, Action<IPacket> callback)
    {
        if (!IsConnected())
        {
            OnError?.Invoke(0, (ushort)ConnectorErrorCode.Disconnected, request);
            return;
        }

        _clientNetwork!.Request(request, callback, 0);
    }

    /// <summary>
    /// 요청 전송 (Stage 없음, async/await 방식)
    /// </summary>
    /// <param name="request">요청 패킷</param>
    /// <returns>응답 패킷</returns>
    public async Task<IPacket> RequestAsync(IPacket request)
    {
        if (!IsConnected())
        {
            throw new ConnectorException(0, (ushort)ConnectorErrorCode.Disconnected, request, 0);
        }

        return await _clientNetwork!.RequestAsync(request, 0);
    }

    #endregion

    #region Send/Request (With Stage)

    /// <summary>
    /// 메시지 전송 (Stage 지정, 응답 없음)
    /// </summary>
    /// <param name="stageId">Stage ID</param>
    /// <param name="packet">전송할 패킷</param>
    public void Send(long stageId, IPacket packet)
    {
        if (!IsConnected())
        {
            OnError?.Invoke(stageId, (ushort)ConnectorErrorCode.Disconnected, packet);
            return;
        }

        _clientNetwork!.Send(packet, stageId);
    }

    /// <summary>
    /// 요청 전송 (Stage 지정, 콜백 방식)
    /// </summary>
    /// <param name="stageId">Stage ID</param>
    /// <param name="request">요청 패킷</param>
    /// <param name="callback">응답 콜백</param>
    public void Request(long stageId, IPacket request, Action<IPacket> callback)
    {
        if (!IsConnected())
        {
            OnError?.Invoke(stageId, (ushort)ConnectorErrorCode.Disconnected, request);
            return;
        }

        _clientNetwork!.Request(request, callback, stageId);
    }

    /// <summary>
    /// 요청 전송 (Stage 지정, async/await 방식)
    /// </summary>
    /// <param name="stageId">Stage ID</param>
    /// <param name="request">요청 패킷</param>
    /// <returns>응답 패킷</returns>
    public async Task<IPacket> RequestAsync(long stageId, IPacket request)
    {
        if (!IsConnected())
        {
            throw new ConnectorException(stageId, (ushort)ConnectorErrorCode.Disconnected, request, 0);
        }

        return await _clientNetwork!.RequestAsync(request, stageId);
    }

    #endregion

    #region Unity Support

    /// <summary>
    /// 메인 스레드에서 콜백 실행 (Unity Update에서 호출)
    /// </summary>
    public void MainThreadAction()
    {
        _clientNetwork?.MainThreadAction();
    }

    #endregion

    #region Cache

    /// <summary>
    /// 캐시 정리
    /// </summary>
    public void ClearCache()
    {
        _clientNetwork?.ClearCache();
    }

    #endregion
}

File: connectors/java/src/main/java/com/playhouse/connector/Connector.java
package com.playhouse.connector;

import com.playhouse.connector.internal.ClientNetwork;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * PlayHouse Connector
 * <p>
 * PlayHouse 실시간 게임 서버에 연결하여 메시지 송수신을 처리합니다.
 * Java NIO 기반 비동기 I/O를 사용하며, 동기/비동기/콜백 세 가지 API 스타일을 지원합니다.
 * <p>
 * <h2>API 스타일</h2>
 * <ul>
 *   <li><b>동기 API</b>: Virtual Thread에서 효율적으로 동작 (Java 21+)</li>
 *   <li><b>비동기 API</b>: CompletableFuture 반환 (*Async 접미사)</li>
 *   <li><b>콜백 API</b>: Consumer/Runnable 콜백 파라미터</li>
 * </ul>
 * <p>
 * <h2>동기 API 사용 예제 (권장 - Virtual Thread)</h2>
 * <pre>
 * ConnectorConfig config = ConnectorConfig.builder()
 *     .requestTimeoutMs(10000)
 *     .build();
 *
 * try (Connector connector = new Connector()) {
 *     connector.init(config);
 *     connector.setOnReceive(packet -> System.out.println("Received: " + packet.getMsgId()));
 *
 *     // Virtual Thread에서 실행
 *     Thread.startVirtualThread(() -> {
 *         try {
 *             // 동기 연결
 *             connector.connect("localhost", 34001);
 *
 *             // 동기 인증
 *             Packet authPacket = Packet.fromBytes("AuthRequest", authData);
 *             boolean authenticated = connector.authenticate(authPacket);
 *
 *             // 동기 요청-응답
 *             Packet request = Packet.fromBytes("EchoRequest", "Hello".getBytes());
 *             Packet response = connector.request(request);
 *             System.out.println("Response: " + response.getMsgId());
 *         } catch (ConnectorException e) {
 *             System.err.println("Error: " + e.getMessage());
 *         }
 *     });
 * }
 * </pre>
 * <p>
 * <h2>비동기 API 사용 예제</h2>
 * <pre>
 * try (Connector connector = new Connector()) {
 *     connector.init();
 *     connector.setOnConnect(() -> System.out.println("Connected"));
 *     connector.setOnReceive(packet -> System.out.println("Received: " + packet.getMsgId()));
 *
 *     connector.connectAsync("localhost", 34001)
 *         .thenCompose(v -> {
 *             Packet authPacket = Packet.fromBytes("AuthRequest", authData);
 *             return connector.authenticateAsync(authPacket);
 *         })
 *         .thenCompose(authenticated -> {
 *             Packet request = Packet.fromBytes("EchoRequest", "Hello".getBytes());
 *             return connector.requestAsync(request);
 *         })
 *         .thenAccept(response -> System.out.println("Response: " + response.getMsgId()))
 *         .join();
 * }
 * </pre>
 * <p>
 * <h2>콜백 API 사용 예제 (게임 엔진용)</h2>
 * <pre>
 * try (Connector connector = new Connector()) {
 *     connector.init();
 *     connector.setOnConnect(() -> System.out.println("Connected"));
 *
 *     // 콜백 방식 요청
 *     Packet request = Packet.fromBytes("EchoRequest", "Hello".getBytes());
 *     connector.request(request, response -> {
 *         System.out.println("Response: " + response.getMsgId());
 *     });
 * }
 * </pre>
 */
public final class Connector implements AutoCloseable {

    private static final Logger logger = LoggerFactory.getLogger(Connector.class);

    private ClientNetwork clientNetwork;
    private ConnectorConfig config;
    private volatile long stageId;
    private volatile boolean initialized;

    /**
     * Connector 초기화
     *
     * @param config 설정
     * @throws IllegalStateException 이미 초기화된 경우
     */
    public void init(ConnectorConfig config) {
        if (initialized) {
            throw new IllegalStateException("Connector already initialized");
        }

        if (config == null) {
            throw new IllegalArgumentException("Config cannot be null");
        }

        this.config = config;
        this.clientNetwork = new ClientNetwork(config);
        this.stageId = 0L;
        this.initialized = true;

        logger.info("Connector initialized: {}", config);
    }

    /**
     * 기본 설정으로 초기화
     */
    public void init() {
        init(ConnectorConfig.defaultConfig());
    }

    /**
     * 서버에 비동기 연결
     *
     * @param host 서버 호스트
     * @param port 서버 포트
     * @return 연결 성공 여부 Future
     * @throws IllegalStateException 초기화되지 않은 경우
     */
    public CompletableFuture<Void> connectAsync(String host, int port) {
        checkInitialized();

        logger.info("Connecting to {}:{}...", host, port);

        return clientNetwork.connectAsync(host, port)
            .thenApply(success -> {
                if (!success) {
                    throw new ConnectorException(1001, "Connection failed");
                }
                return null;
            });
    }

    /**
     * 서버에 동기 연결 (Virtual Thread에서 효율적으로 동작)
     *
     * @param host 서버 호스트
     * @param port 서버 포트
     * @throws ConnectorException    연결 실패 시
     * @throws IllegalStateException 초기화되지 않은 경우
     */
    public void connect(String host, int port) {
        try {
            connectAsync(host, port).join();
        } catch (CompletionException e) {
            throw unwrapException(e);
        }
    }

    /**
     * 연결 해제
     */
    public void disconnect() {
        if (!initialized) {
            return;
        }

        logger.info("Disconnecting...");
        clientNetwork.disconnectAsync().join();
    }

    /**
     * 연결 상태 확인
     *
     * @return 연결되어 있으면 true
     */
    public boolean isConnected() {
        return initialized && clientNetwork.isConnected();
    }

    /**
     * 인증 상태 확인
     *
     * @return 인증되어 있으면 true
     */
    public boolean isAuthenticated() {
        return initialized && clientNetwork.isAuthenticated();
    }

    /**
     * 인증 요청 (간편 API - 비동기)
     *
     * @param serviceId 서비스 ID
     * @param accountId 계정 ID
     * @param payload   인증 페이로드
     * @return 인증 성공 여부 Future
     */
    public CompletableFuture<Boolean> authenticateAsync(String serviceId, String accountId, byte[] payload) {
        // 실제 구현에서는 AuthRequest 프로토콜 사용
        Packet authPacket = Packet.builder("AuthRequest")
            .payload(payload)
            .build();

        return requestAsync(authPacket)
            .thenApply(response -> {
                boolean success = response.getErrorCode() == 0;
                if (success) {
                    logger.info("Authentication successful: serviceId={}, accountId={}", serviceId, accountId);
                } else {
                    logger.warn("Authentication failed: errorCode={}", response.getErrorCode());
                }
                return success;
            });
    }

    /**
     * 인증 요청 (동기 방식 - Virtual Thread에서 효율적으로 동작)
     *
     * @param serviceId 서비스 ID
     * @param accountId 계정 ID
     * @param payload   인증 페이로드
     * @return 인증 성공 여부
     * @throws ConnectorException 인증 실패 시
     */
    public boolean authenticate(String serviceId, String accountId, byte[] payload) {
        try {
            return authenticateAsync(serviceId, accountId, payload).join();
        } catch (CompletionException e) {
            throw unwrapException(e);
        }
    }

    /**
     * Packet을 사용한 인증 요청 (비동기)
     *
     * @param authPacket 인증 패킷
     * @return 인증 성공 여부 Future
     */
    public CompletableFuture<Boolean> authenticateAsync(Packet authPacket) {
        return requestAsync(authPacket)
            .thenApply(response -> {
                boolean success = response.getErrorCode() == 0;
                if (success) {
                    logger.info("Authentication successful");
                } else {
                    logger.warn("Authentication failed: errorCode={}", response.getErrorCode());
                }
                return success;
            });
    }

    /**
     * Packet을 사용한 인증 요청 (동기 방식 - Virtual Thread에서 효율적으로 동작)
     *
     * @param authPacket 인증 패킷
     * @return 인증 성공 여부
     * @throws ConnectorException 인증 실패 시
     */
    public boolean authenticate(Packet authPacket) {
        try {
            return authenticateAsync(authPacket).join();
        } catch (CompletionException e) {
            throw unwrapException(e);
        }
    }


    /**
     * 메시지 전송 (응답 없음)
     *
     * @param packet 전송할 패킷
     * @throws IllegalStateException 초기화되지 않았거나 연결되지 않은 경우
     */
    public void send(Packet packet) {
        checkConnected();
        clientNetwork.send(packet, stageId);
    }

    /**
     * 요청 전송 (비동기 응답)
     *
     * @param packet 요청 패킷
     * @return 응답 패킷 Future
     * @throws IllegalStateException 초기화되지 않았거나 연결되지 않은 경우
     */
    public CompletableFuture<Packet> requestAsync(Packet packet) {
        checkConnected();
        return clientNetwork.requestAsync(packet, stageId);
    }

    /**
     * 요청 전송 (동기 방식 - Virtual Thread에서 효율적으로 동작)
     *
     * @param packet 요청 패킷
     * @return 응답 패킷
     * @throws ConnectorException    요청 실패 시
     * @throws IllegalStateException 초기화되지 않았거나 연결되지 않은 경우
     */
    public Packet request(Packet packet) {
        try {
            return requestAsync(packet).join();
        } catch (CompletionException e) {
            throw unwrapException(e);
        }
    }

    /**
     * 요청 전송 (콜백 방식)
     *
     * @param packet   요청 패킷
     * @param callback 응답 콜백
     * @throws IllegalStateException 초기화되지 않았거나 연결되지 않은 경우
     */
    public void request(Packet packet, Consumer<Packet> callback) {
        requestAsync(packet)
            .thenAccept(callback)
            .exceptionally(e -> {
                logger.error("Request failed: {}", packet.getMsgId(), e);
                return null;
            });
    }

    /**
     * 메인 스레드에서 콜백 실행
     * <p>
     * Unity, Godot 등의 게임 엔진에서 메인 스레드에서 호출해야 합니다.
     * 일반 서버 애플리케이션에서는 필요하지 않습니다.
     */
    public void mainThreadAction() {
        if (initialized) {
            clientNetwork.mainThreadAction();
        }
    }

    /**
     * Stage ID 설정
     *
     * @param stageId Stage ID
     */
    public void setStageId(long stageId) {
        this.stageId = stageId;
    }

    /**
     * 현재 Stage ID 반환
     *
     * @return Stage ID
     */
    public long getStageId() {
        return stageId;
    }

    // ===== 콜백 설정 =====

    /**
     * 연결 콜백 설정
     *
     * @param callback 연결 결과 콜백 (성공 여부)
     */
    public void setOnConnect(Runnable callback) {
        checkInitialized();
        clientNetwork.setOnConnect(success -> callback.run());
    }

    /**
     * 메시지 수신 콜백 설정
     *
     * @param callback 수신 콜백
     */
    public void setOnReceive(Consumer<Packet> callback) {
        checkInitialized();
        clientNetwork.setOnReceive(callback);
    }

    /**
     * 에러 콜백 설정
     *
     * @param callback 에러 콜백 (에러 코드, 메시지)
     */
    public void setOnError(BiConsumer<Integer, String> callback) {
        checkInitialized();
        clientNetwork.setOnError(callback);
    }

    /**
     * 연결 해제 콜백 설정
     *
     * @param callback 연결 해제 콜백
     */
    public void setOnDisconnect(Runnable callback) {
        checkInitialized();
        clientNetwork.setOnDisconnect(callback);
    }

    // ===== AutoCloseable =====

    /**
     * 리소스 정리
     */
    @Override
    public void close() {
        if (initialized) {
            logger.info("Closing connector");
            disconnect();
            clientNetwork.shutdown();
            initialized = false;
        }
    }

    // ===== Private Methods =====

    private void checkInitialized() {
        if (!initialized) {
            throw new IllegalStateException("Connector not initialized. Call init() first.");
        }
    }

    private void checkConnected() {
        checkInitialized();
        if (!isConnected()) {
            throw new IllegalStateException("Not connected to server");
        }
    }

    /**
     * CompletionException을 unwrap하여 원본 예외로 변환
     * ConnectorException이 아닌 경우 새로운 ConnectorException으로 감싼다.
     *
     * @param e CompletionException
     * @return 원본 RuntimeException 또는 ConnectorException
     */
    private RuntimeException unwrapException(CompletionException e) {
        Throwable cause = e.getCause();

        // ConnectorException인 경우 그대로 반환
        if (cause instanceof ConnectorException connectorEx) {
            return connectorEx;
        }

        // RuntimeException인 경우 그대로 반환
        if (cause instanceof RuntimeException runtimeEx) {
            return runtimeEx;
        }

        // 그 외의 경우 ConnectorException으로 감싸서 반환
        return new ConnectorException(
            5000,
            "Unexpected error: " + cause.getMessage(),
            cause
        );
    }

    /**
     * 현재 설정 반환
     *
     * @return ConnectorConfig
     */
    public ConnectorConfig getConfig() {
        return config;
    }
}
